### PR TITLE
Make Empty Map Diffs more Efficient

### PR DESCRIPTION
**Pull request by original-brownbear**
Use empty singletons in a few more places and make empty diffs noops so that
diff application does not result in redundant map copying in `#apply` (also making comparisons faster all over the place because we get instance equality in more cases).

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java [hunk 1]
```diff
@@ -65,47 +65,59 @@ public void writeTo(StreamOutput out) throws IOException {
...
-                }
-            }
-            deletes = tempDeletes;
-            upserts = tempUpserts;
-        }
-
-        private DiffableStringMapDiff(StreamInput in) throws IOException {
-            deletes = in.readStringList();
-            upserts = in.readMap(StreamInput::readString, StreamInput::readString);
+        private DiffableStringMapDiff(final List<String> deletes, final Map<String, String> upserts) {
```
Comment (arteam) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java:118]: `final` seems to redundant for both parameters.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java [hunk 2]
```diff
@@ -65,47 +65,59 @@ public void writeTo(StreamOutput out) throws IOException {
 
     @Override
     public Diff<DiffableStringMap> diff(DiffableStringMap previousState) {
-        return new DiffableStringMapDiff(previousState, this);
+        if (this.equals(previousState)) {
```
Comment (arteam) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java:68]: `this.` seem superfluous here

File: server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java [hunk 3]
```diff
@@ -65,47 +65,59 @@ public void writeTo(StreamOutput out) throws IOException {
...
+            return DiffableStringMapDiff.EMPTY;
+        }
+        final List<String> tempDeletes = new ArrayList<>();
+        final Map<String, String> tempUpserts = new HashMap<>();
```
Comment (arteam) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java:72]: I think it's better to declare `tempUpserts` before the 2nd loop where it's actually used.

Comment (arteam) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java:72]: I know the code part was just moved, but I still think it makes sense to improve it a bit.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java [hunk 4]
```diff
@@ -65,47 +65,59 @@ public void writeTo(StreamOutput out) throws IOException {
...
+        return new DiffableStringMapDiff(deletes, upserts);
     }
 
     /**
      * Represents differences between two DiffableStringMaps.
      */
     public static class DiffableStringMapDiff implements Diff<DiffableStringMap> {
 
-        public static final DiffableStringMapDiff EMPTY = new DiffableStringMapDiff(DiffableStringMap.EMPTY, DiffableStringMap.EMPTY);
+        public static final DiffableStringMapDiff EMPTY = new DiffableStringMapDiff(Collections.emptyList(), Collections.emptyMap()) {
```
Comment (arteam) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DiffableStringMap.java:108]: Nit: For consistency sake I think makes sense to use either (`List.of()` and `Map.of`) or  (`Collections.emptyList()` and `Collections.emptyMap()`) both in `DiffableUtils` and  `DiffableStringMap`.

**Review by arteam — APPROVED** [review]
Thanks Armin. LGTM, I've left a couple of small comments.

**Review by arteam — APPROVED** [review]

**Comment by original-brownbear** [comment]
Thanks Artem!

**Pull request merged**